Sources/Brave/Frontend/Browser/Interstitial Pages/BraveSchemeHandler.swift

```diff
@@ -0,0 +1,106 @@
+import WebKit
+import Shared
+
+class BraveSchemeHandler: NSObject, WKURLSchemeHandler {
+  static let scheme = "brave"
+  
+  enum Host: String, CaseIterable {
+    case player
+    
+    /// The way the navigation bar should display this value
+    var navigationBarDisplayString: String {
+      return [BraveSchemeHandler.scheme, rawValue].joined(separator: "://")
+    }
+  }
+  
+  typealias ResourceInfo = (contentType: String, fileName: String)
+  
+  /// A mapping of a path to mime type
+  private let mimeTypeMapping: [String: String] = [
+    // Brave Player
+    "/styles/Player.css": "text/css",
+    "/scripts/Player.js": "text/javascript",
+    "/images/PlayerBackground.svg": "image/svg+xml",
+    "/images/PlayerLogo.svg": "image/svg+xml"
+  ]
+  
+  func webView(_ webView: WKWebView, start urlSchemeTask: WKURLSchemeTask) {
+    guard let url = urlSchemeTask.request.url, let host = Self.host(for: url) else {
+      urlSchemeTask.didFailWithError(InternalPageSchemeHandlerError.notAuthorized)
+      return
+    }
+    
+    do {
+      if let resource = try resource(for: url) {
+        let response = URLResponse(url: url, mimeType: resource.mimeType, expectedContentLength: -1, textEncodingName: nil)
+        urlSchemeTask.didReceive(response)
+        urlSchemeTask.didReceive(resource.data)
+        urlSchemeTask.didFinish()
+      } else if let data = try host.loadData(for: url) {
+        let response = URLResponse(url: url, mimeType: "text/html", expectedContentLength: -1, textEncodingName: "utf-8")
+        urlSchemeTask.didReceive(response)
+        urlSchemeTask.didReceive(data)
+        urlSchemeTask.didFinish()
+      } else {
+        urlSchemeTask.didFailWithError(InternalPageSchemeHandlerError.notAuthorized)
+      }
+    } catch {
+      urlSchemeTask.didFailWithError(error)
+    }
+  }
+  
+  func webView(_ webView: WKWebView, stop urlSchemeTask: WKURLSchemeTask) {
+    // Not handled
+  }
+  
+  /// Checks to see if this scheme handler handles this request. This is needed for the navigation delegate
+  static func handles(url: URL) -> Bool {
+    return host(for: url) != nil
+  }
+  
+  private func resource(for url: URL) throws -> (data: Data, mimeType: String)? {
+    let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+    guard let path = components?.path else { return nil }
+    guard let mimeType = mimeTypeMapping[path] else { return nil }
+    
+    guard let resourceURL = Bundle.module.url(forResource: url.lastPathComponent, withExtension: nil) else {
+      return nil
+    }
+    
+    let data = try Data(contentsOf: resourceURL)
+    return (data, mimeType)
+  }
+  
+  static func host(for url: URL) -> Host? {
+    let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+    guard components?.scheme == scheme else { return nil }
+    guard let host = components?.host else { return nil }
+    return Host(rawValue: host)
+  }
+}
+
+extension BraveSchemeHandler.Host {
```

Comment:
what happens if an external page or app tries to open a `brave://` URL? note on desktop that navigations to `brave://` or embeddings of `brave://` are blocked for security reasons.

Reply:
Can confirm a webpage currently can open brave://player. 
```javascript
      var a = document.createElement('a');
      var linkText = document.createTextNode("player");
      a.appendChild(linkText);
      a.id = "player"
      a.href = "brave://player/youtube.com/EKevmQMgpQs";
      document.body.appendChild(a);
      document.getElementById('player').click()
```
As @diracdeltas said, that should be blocked in a similar way it's blocked for internal://local URLs
https://github.com/brave/brave-ios/blob/705bb485a8fc4267675259a43dc9df8cc096b476/Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController%2BWKNavigationDelegate.swift#L146

Reply:
I can satisfy the following requirements without much effort:
1. Restrict brave:// scheme requests in subframes
2. Restrict brave:// scheme requests with unsupported hosts (e.x. `brave://fake`)
3. Restring link navigations.

What I'm not totally sure about is:
1. Restrict other type of navigations such as window.location.apply

For the above, I might be able to limit it by applying a uuid and checking it during navigation. I now understand why it was there for `internal://` navigations. However this will likely have the following consequences:
1. Im not sure if I can allow users to manually enter links to brave://player in the url bar. This was never needed for internal:// links so this was never a problem. I might be able to overcome this with some trickery but this will take some further investigation

Reply:
Neither `window.location = 'brave://player'` nor `window.open('brave://player')` work for me in the current implementation, so maybe WebKit already prevents such navigations.

Reply:
My bad, `window.location = 'brave://player'` actually works, but `window.open('brave://player')` doesn't

Reply:
Yah I can't get window.open to work either. I managed to secure `window.location` The consequences of this however are a lost back-forth history when restoring tabs. I personally think this is acceptable for now. I hope others agree.

Reply:
I created a jsFiddle [here](https://jsfiddle.net/twofox/s4w01hn7/7/). but jsFiddle will be blocked for all navigations simply because it runs in a iframe so it's not a good test for the above (i.e. the tests will pass but because of main frame restrictions, not anything else). 

I tested by temporarily removing frame restrictions in code so auth restrictions kick in. But maybe we should create a test page for these. Maybe @pes10k can help here?

Reply:
you can make a test page by pushing index.html to a new github repo (personal is fine) and enabling github pages.

Reply:
@cuba sure happy to help put something up on https://dev-pages.brave.software

Reply:
@pes10k As we discussed I should be able to write an automated test for this particular case. 
@diracdeltas Yah ok I can GitHub page might be useful for other things in the long run.